### DISCUSSION TITLE
Error of in-flight requests using streaming response

### DISCUSSION THREAD

**Discussion by gtsystem** [encode/httpx#2118] [Potential Issue] [answered]
Hi, I'm trying to debug an issue that started showing up from httpx 0.20.
I created this mini example that is exposing the problem:

```python
import json
import asyncio
import httpx

async def read_lines(client: httpx.AsyncClient):
    async with client.stream("GET", "https://httpbin.org/stream/10") as resp:
        async for line in resp.aiter_lines():
            yield json.loads(line)

async def main():
    client = httpx.AsyncClient()
    async for json_line in read_lines(client):
        print(json_line)
        break   # we interrupt reading the stream

    await client.aclose()

asyncio.run(main())
```

And it fails with

    RuntimeError: The connection pool was closed while 1 HTTP requests/responses were still in-flight.

From my investigation it seems that `resp.aclose()` is called first (inside the context), but as it is asyncronous, before completing, the code reach `async client.aclose()` and the error is raised.

Adding `await asyncio.sleep(0.1)` before `async client.aclose()` resolve the issue.

Is it a bug or I'm doing something wrong? Any idea?

**Comment by guyskk** [comment, marked as answer]
Hi @gtsystem, it seems an asynchronous generators finalization problem, FYI: https://peps.python.org/pep-0525/#finalization

Close the generator manually can resolve the problem.

```python
async def main():
    client = httpx.AsyncClient()
    agen = read_lines(client)
    async for json_line in agen:
        print(json_line)
        break  # we interrupt reading the stream
    await agen.aclose()  # --> close it manually
    await client.aclose()
```

**Reply by gtsystem** [reply]
Yes, it looks like this is solving the issue for me. Thanks!
